src/renderers/shared/stack/reconciler/ReactCompositeComponent.js

```diff
@@ -1100,10 +1100,17 @@ var ReactCompositeComponent = {
    * @final
    * @private
    */
-  attachRef: function(ref, component) {
+  attachRef: function(ref, component, mockConfig) {
     var inst = this.getPublicInstance();
     invariant(inst != null, 'Stateless function components cannot have refs.');
-    var publicComponentInstance = component.getPublicInstance();
+    var publicComponentInstance;
+    if (mockConfig && mockConfig.getMockRef) {
```

Comment:
@gaearon one issue is that `component.getPublicInstance()` will be calling `ReactCompositeComponent.getPublicInstance` if the component is a composite. Even if I pass `transaction` all the way to `ReactCompositeComponent`, it would have to be aware of `transaction.mockConfig` if we wanted composite components to return mock instances.